File: 0503-next-greater-element-ii/0503-next-greater-element-ii.cpp
class Solution {
public:
    vector<int> nextGreaterElements(vector<int>& nums) {
       int n = nums.size();
       stack<int> s;
       vector<int> res(n, -1);
        
     for(int i = 2 * n - 1; i >= 0; --i) {
      int num = nums[i % n];
     while(!s.empty() && s.top() <= num) s.pop(); // remove the small one(last onr in stack)
     if(!s.empty()) res[i % n] = s.top(); // add into vector the top element
     s.push(num); 
     }
     return res;
    }
};

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int n = nums.length;
        Stack<Integer>st = new Stack<>();
        int[] result = new int[n];
        Arrays.fill(result, -1);
        
        for(int i = 0; i < 2*n; i++){
            int curr_idx = i % n;

            while(!st.isEmpty() && nums[curr_idx] > nums[st.peek()]){
                int idx = st.pop();
                result[idx] = nums[curr_idx];
            }
            if(i < n){
                st.push(curr_idx);
            }
        }
        return result;
    }
}